### PR TITLE
Support rapid hinting mode in yanking

### PR DESCRIPTION
**Pull request by rr-**
Implements #3824 

I took care of clearing clipboard on first yank by adding `num` to `HintContext`.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3825)

### CODE REVIEW

File: qutebrowser/browser/hints.py [hunk 1]
```diff
@@ -240,7 +243,18 @@ def yank(self, url, context):
...
+        new_content = urlstr
+
+        # only second and consecutive yanks are to append to the clipboard
+        if context.rapid and context.num > 1:
+            try:
+                old_content = utils.get_clipboard(selection=sel)
+            except utils.ClipboardEmptyError:
+                pass
+            else:
+                new_content = old_content + '\n' + new_content
```
Comment (toofar) [inline qutebrowser/browser/hints.py]: I think this should be `os.linesep.join(old_content, new_content)` for our windows and mac using friends.

File: qutebrowser/browser/hints.py [hunk 2]
```diff
@@ -240,7 +243,18 @@ def yank(self, url, context):
...
-        utils.set_clipboard(urlstr, selection=sel)
+
+        new_content = urlstr
+
+        # only second and consecutive yanks are to append to the clipboard
+        if context.rapid and context.num > 1:
+            try:
+                old_content = utils.get_clipboard(selection=sel)
+            except utils.ClipboardEmptyError:
+                pass
```
[the remark this answers is not in the document]
Comment (rr-) [reply qutebrowser/browser/hints.py]: But that's how it works.
If the clipboard is empty, line 258 still gets executed and writes the first yanked link to the clipboard.

[the remark this answers is not in the document]
Comment (jgkamat) [reply qutebrowser/browser/hints.py]: ah, never mind then, ignore me :)

**Comment by The-Compiler** [comment]
Just a quick thought without much review of the code: The `num` thing seems more complicated than necessary :wink: Can't we just clear the clipboard once when entering the hint mode? Or if not, wouldn't a `first_run` boolean or so suffice?

**Comment by rr-** [comment]
>Can't we just clear the clipboard once when entering the hint mode? 

I thought about this, but TBH I haven't managed to find the place where I could do that.

>Or if not, wouldn't a first_run boolean or so suffice?

Yep, it would.

Please let me know which approach you like the best and I'll adjust the code accordingly.

**Comment by jgkamat** [comment]
I would personally prefer the `first_run`, I don't think we would ever use `num`, and here we're only using it as a `first_run` anyway

**Comment by rr-** [comment]
Changed `num` to `first_run`.

**Comment by rr-** [comment]
Unfortunately I can't get the tests to run, but I've added one nonetheless. Let's see what travis says about it.

**Comment by rr-** [comment]
The tests have failed:
```
    @bdd.then(bdd.parsers.re(r'the (?P<what>primary selection|clipboard) should '
                             r'contain "(?P<content>.*)"'))
    def clipboard_contains(quteproc, server, what, content):
        expected = content.replace('(port)', str(server.port))
        expected = expected.replace('\\n', '\n')
        quteproc.wait_for(message='Setting fake {}: {}'.format(
>           what, json.dumps(expected)))
E       end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000.0ms waiting for {'message': 'Setting fake clipboard: "http://localhost:1059/data/hello.txt\\nhttp://localhost:1059/data/hello2.txt"'}.

tests\end2end\features\conftest.py:604: WaitForTimeout
```
Maybe it has to do with `\n`?

**Comment by toofar** [comment]
The test is failing because `debug-set-fake-clipboard` makes `util.set_clipboard()` not actually set anything, just print out that it does.  It should also be sitting `fake_clipboard`. Try it with this:
```
diff --git i/qutebrowser/utils/utils.py w/qutebrowser/utils/utils.py
index da1ddf085127..4554aef2e920 100644
--- i/qutebrowser/utils/utils.py
+++ w/qutebrowser/utils/utils.py
@@ -510,11 +510,13 @@ def sanitize_filename(name, replacement='_'):
 
 def set_clipboard(data, selection=False):
     """Set the clipboard to some given data."""
+    global fake_clipboard
     if selection and not supports_selection():
         raise SelectionUnsupportedError
     if log_clipboard:
         what = 'primary selection' if selection else 'clipboard'
         log.misc.debug("Setting fake {}: {}".format(what, json.dumps(data)))
+        fake_clipboard = data
     else:
         mode = QClipboard.Selection if selection else QClipboard.Clipboard
         QApplication.clipboard().setText(data, mode=mode)
```

**Comment by rr-** [comment]
After the change, the tests fail with:

```
_________________________ TestGetSetClipboard.test_get _________________________
self = <test_utils.TestGetSetClipboard object at 0x7f981befeb38>
    def test_get(self):
>       assert utils.get_clipboard() == 'mocked clipboard text'
E       AssertionError: assert 'füb' == 'mocked clipboard text'
E         - füb
E         + mocked clipboard text
tests/unit/utils/test_utils.py:681: AssertionError
------------------------------ Captured log call -------------------------------
utils.py                   553 DEBUG    Clipboard contained: 'füb'
```

**Comment by toofar** [comment]
Drat, I suspect those tests will need to be switched from using
`debug-set-fake-clipboard` to using a proper mock like the one in
`test_utils.py`.

**Comment by toofar** [comment]
Well to stop `TestGetSetClipboard.test_get()` from failing something like the following works:
```
diff --git i/tests/unit/utils/test_utils.py w/tests/unit/utils/test_utils.py
index df0eb9ecb0ea..ef2b6c8d4832 100644
--- i/tests/unit/utils/test_utils.py
+++ w/tests/unit/utils/test_utils.py
@@ -651,6 +651,7 @@ class TestGetSetClipboard:
                          autospec=True)
         clipboard = m()
         clipboard.text.return_value = 'mocked clipboard text'
+        mocker.patch('qutebrowser.utils.utils.fake_clipboard', None)
         return clipboard
 
     def test_set(self, clipboard_mock, caplog):

```
But I still don't get why this `:debug-set-fake-clipboard` thing exists and why it is being used in the end2end tests instead of a proper mock.

**Comment by The-Compiler** [comment]
> But I still don't get why this `:debug-set-fake-clipboard` thing exists and why it is being used in the end2end tests instead of a proper fixture/mock.

Because in end2end tests, qutebrowser runs in a subprocess, so you can't mock things in it. I want to get rid of most of those tests in the long run, though.

**Comment by rr-** [comment]
More errors ._·

**Comment by toofar** [comment]
Alright, this time it is about the line endings. Try this patch and cross your fingers, we can do this!
```
diff --git i/tests/end2end/features/conftest.py w/tests/end2end/features/conftest.py
index 4752e2393328..d81c40b2a057 100644
--- i/tests/end2end/features/conftest.py
+++ w/tests/end2end/features/conftest.py
@@ -599,7 +599,7 @@ def check_open_tabs(quteproc, request, tabs):
                          r'contain "(?P<content>.*)"'))
 def clipboard_contains(quteproc, server, what, content):
     expected = content.replace('(port)', str(server.port))
-    expected = expected.replace('\\n', '\n')
+    expected = expected.replace('\\n', os.linesep)
     quteproc.wait_for(message='Setting fake {}: {}'.format(
         what, json.dumps(expected)))
 
diff --git i/tests/end2end/features/hints.feature w/tests/end2end/features/hints.feature
index e7642450dd9c..2399fb37201c 100644
--- i/tests/end2end/features/hints.feature
+++ w/tests/end2end/features/hints.feature
@@ -127,7 +127,7 @@ Feature: Using hints
         And I run :follow-hint a
         And I run :follow-hint s
         And I run :leave-mode
-        Then the clipboard should contain "http://localhost:(port)/data/hello.txt\r\nhttp://localhost:(port)/data/hello2.txt"
+        Then the clipboard should contain "http://localhost:(port)/data/hello.txt\nhttp://localhost:(port)/data/hello2.txt"
 
     Scenario: Rapid hinting
         When I open data/hints/rapid.html in a new tab
```

**Comment by toofar** [comment]
Oh I thought those caret ones would fail then saw that the mac one was passing, guess I missed windows. I don't know why the caret stuff is putting `\n`s in the clipboard on windows or if that is intended or an oversight. If it is intended I think the only way to get both working is to make a new bdd `then` action, I would like to be able to pass a list of strings from the scenario and have the test construct the expected string from that but I don't think that is possible.

**Comment by toofar** [comment]
The `\n` in the caret-selected text just comes from `window.getSelection().ToString()`. Apparently what `<br/>` gets turned into for that is not defined in the spec and implementations differ (from 2015): https://github.com/w3c/selection-api/issues/7
I think [TextIteratorAlgorithm\<Strategy>::handleNonTextNode()](https://chromium.googlesource.com/chromium/src/+/dc57b581dc8fd4efef4a7ca42e0d28f924e0e629/third_party/WebKit/Source/core/editing/iterators/TextIterator.cpp#938) is where chrome does this stuff and it does indeed seem to hardcode a bare `\n`. Interesting.
So if we want platform dependant line breaks in the yanked string we can either not use the same checking function as the javascript stuff or do something like:
```
diff --git i/tests/end2end/features/conftest.py w/tests/end2end/features/conftest.py
index 4752e2393328..d81c40b2a057 100644
--- i/tests/end2end/features/conftest.py
+++ w/tests/end2end/features/conftest.py
@@ -599,7 +599,7 @@ def check_open_tabs(quteproc, request, tabs):
                          r'contain "(?P<content>.*)"'))
 def clipboard_contains(quteproc, server, what, content):
     expected = content.replace('(port)', str(server.port))
     expected = expected.replace('\\n', '\n')
+    expected = expected.replace('(linesep)', os.linesep)
     quteproc.wait_for(message='Setting fake {}: {}'.format(
         what, json.dumps(expected)))
```
And change the new test to use `(linesep)` instead of `\n`.

**Comment by The-Compiler** [comment]
I feel like we should do a `.replace('\n', os.linesep)` in the caret browsing code, so we actually get CRLF on Windows (like a Windows user would probably expect). I can take a look at that.

**Comment by toofar** [comment]
Okay but this is only for eg `<br/>` tags, what if you are yanking code
that legitimately has bare newlines in it on windows? I just checked
with
https://patch-diff.githubusercontent.com/raw/qutebrowser/qutebrowser/pull/3825.patch
on windows (using vivaldi) and it didn't have windows line endings.

**Comment by toofar** [comment]
Can you clarify what needs to be done to get this one over the line? Currently the tests are breaking on windows because I suggested to make the `clipboard_contains` test function expect os-dependant line endings, but the caret code which also uses that function is, in some cases, returning `\n` on all platforms. In your last comment you suggested changing the caret code to return os-dependant line endings "like a Windows user would probably expect", but given that other chromium based browsers on windows don't do that I don't think that's true. Also the problem with chrome adding `\n`s in only for when it converts `</br>` elemnts to text, everything else is passed through as it is in the copied text and we probably shouldn't change that.

So unless you actually want to change the caret code to convert line endings I think this patch should be sufficient to get the tests working again, it just reverts a previous change and adds `(linesep)` as a marker for os-dependant line endings.

```diff
diff --git i/tests/end2end/features/conftest.py w/tests/end2end/features/conftest.py
index d81c40b2a057..8040d84cc51b 100644
--- i/tests/end2end/features/conftest.py
+++ w/tests/end2end/features/conftest.py
@@ -599,7 +599,8 @@ def check_open_tabs(quteproc, request, tabs):
                          r'contain "(?P<content>.*)"'))
 def clipboard_contains(quteproc, server, what, content):
     expected = content.replace('(port)', str(server.port))
-    expected = expected.replace('\\n', os.linesep)
+    expected = expected.replace('\\n', '\n')
+    expected = expected.replace('(linesep)', os.linesep)
     quteproc.wait_for(message='Setting fake {}: {}'.format(
         what, json.dumps(expected)))
 
diff --git i/tests/end2end/features/hints.feature w/tests/end2end/features/hints.feature
index 2399fb37201c..f31094bc27ab 100644
--- i/tests/end2end/features/hints.feature
+++ w/tests/end2end/features/hints.feature
@@ -127,7 +127,7 @@ Feature: Using hints
         And I run :follow-hint a
         And I run :follow-hint s
         And I run :leave-mode
-        Then the clipboard should contain "http://localhost:(port)/data/hello.txt\nhttp://localhost:(port)/data/hello2.txt"
+        Then the clipboard should contain "http://localhost:(port)/data/hello.txt(linesep)http://localhost:(port)/data/hello2.txt"
 
     Scenario: Rapid hinting
         When I open data/hints/rapid.html in a new tab
```

**Comment by The-Compiler** [comment]
I agree only fixing the tests is the right thing to do. I can take a closer look probably later today, planning to look at all open PRs over the next few days.

**Comment by The-Compiler** [comment]
Looks good with that - thanks @rr-, @jgkamat and @toofar! :+1:

**Pull request merged**